Automattic/woocommerce-payments
# Integrate split UPE payments with WooCommerce Blocks

Fixes https://github.com/Automattic/woocommerce-payments/issues/4603

#### Changes proposed in this Pull Request
Block checkout registers UPE as a payment method (Introduced by https://github.com/Automattic/woocommerce-payments/pull/2429). What this PR attempts to do is to register multiple payment methods instead.

#### Difference in payment methods management in classic checkout vs Blocks 🧐
Classic checkout renders and manages individual payment methods via WC Core and [`woocommerce_payment_gateways` filter](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments.php#L365).

WooCommerce Blocks provide the `PaymentMethodRegistry` for the method to be registered. During the payment method registration, we [rollout js scripts and styles](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-upe-blocks-payment-method.php#L17), and this time [the side responsible for the methods rendering is the front-end](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/blocks/upe.js#L43) where blocks give us an assist in the form of [registerPaymentMethod](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks-registry/payment-methods/registry.ts#L34-L53).

#### Problem solution 👨🏽‍💻
Having in mind the above, I took the opportunity and used the [`paymentMethodsConfig` field](https://github.com/Automattic/woocommerce-payments/blob/53e37c973707d8b084153655c4d393439c1ec091/includes/payment-methods/class-upe-payment-gateway.php#L717), which comes from the back-end and is exposed in JavaScript. This field represents all the enabled payment methods from WCPay settings. On the front end, [I take all the enabled payment methods and register a payment method for each of them](https://github.com/Automattic/woocommerce-payments/pull/4681/files#diff-f6bacead8106193d743286ebb7f58e7bad5b05526103306750625a01943cebb5R65). For each payment method, [its name is being provided to the component](https://github.com/Automattic/woocommerce-payments/pull/4681/files#diff-f6bacead8106193d743286ebb7f58e7bad5b05526103306750625a01943cebb5R70) in order to be used later on in the [effect hook to create a payment intent](https://github.com/Automattic/woocommerce-payments/blob/53e37c973707d8b084153655c4d393439c1ec091/client/checkout/blocks/upe-fields.js#L370-L408), right after opening the payment method on the UI.  

I also [provide testingInstructions](https://github.com/Automattic/woocommerce-payments/pull/4681/files#diff-f6bacead8106193d743286ebb7f58e7bad5b05526103306750625a01943cebb5R72) for each particular payment method to be displayed on the UI, if we are in the Test mode.

#### How the basic Card Element replaced the card Stripe payment element 
The idea of replacing the card payment method with the legacy card is pretty straightforward - [ignore the card payment method while attaching all the enabled UPE payment methods](https://github.com/Automattic/woocommerce-payments/pull/4681/files#diff-b541c0a287f40f82a1d0e43c52c7c7a2b67ae233f261655cbb437726fec83feeR1071-R1073) and attach the legacy card separately. The same idea is applied in the classic split UPE checkout. My first thought was to update the [UPE implementation](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/blocks/upe.js) to somehow be able to inject a legacy card there. But then I thought - we already have [the fully working separate component with the working legacy card](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/blocks/index.js), which is used when UPE aren't enabled.

Finally, what I did is I rolled out two components [here](https://github.com/Automattic/woocommerce-payments/pull/4681/files#diff-c5f750f16fa665d48bb6ccb2b37c5d3f10d80e18158ce7419555aa3ffd1cf0e6R17-R29) - the legacy card and UPE payment elements without the card. This approach helped us to reduce the amount of global refactoring and allowed us to roll back the solution very quickly in case we will ever need to take a new path.

#### Testing instructions
1. Go to the `wp-admin` console and create a new page. Search for the checkout block and add it to the page. 
![image](https://user-images.githubusercontent.com/572862/193883480-99997c74-6eba-4714-8762-ccfa2b9118fe.png)
2. Go add some items to your cart.
3. Go to your new block checkout page (ie. http://localhost:8082/block-checkout/)
4. You should see all the enabled UPE payment methods render.

<img width="631" alt="image" src="https://user-images.githubusercontent.com/22319444/198616773-9f072e34-806c-4efd-8cbb-014d89f5f778.png">

5. Go to `Payments` -> `Settings` and disable some of the payment methods from the list. Then, go to the blocks checkout page and confirm, that the rendered methods list reflects only the enabled payment methods. 
6. Confirm, that there are test messages appearing for Card and SEPA payment.
7. Make sure the confirmed order is on the Orders list in your WooCommerce console in the admin account. Check transaction details.

<img width="1918" alt="image" src="https://user-images.githubusercontent.com/22319444/198610298-fdc0b59b-365d-4e4a-8858-261e73303256.png">
<img width="1292" alt="image" src="https://user-images.githubusercontent.com/22319444/198614907-05705540-9fca-4532-a619-0b4a1062cb90.png">

8. Perform regression tests with non-UPE checkout and confirm, that card payment works as expected.



-------------------


**Post merge**
